### DISCUSSION TITLE
How to make the shape of the pores more complex？

### DISCUSSION THREAD

**Discussion by Xiaozizhuo** [PMEAL/OpenPNM#2415] [Q&A]
Hello dear professor jeff, in the framework of OpenPNM, can it be more complicated besides generating a simplified model of "the throat is a cylinder and the pores are a sphere"? For example, the pores are cuboid or the porous medium is composed of a large number of cylinders stacked like a GDL.

**Comment by jgostick** [comment]
Hi @Xiaozizhuo 
Which version of openpnm are you using? What options are available to you under ``op.geometry``?  Do you see ``Cubes_and_Cuboids``?  We have several pre-written geometry classes that are slighly more complex.

If you really want full control then you may want to consider writing your own functions.  For instance, if you want the throats to be triangular, you could write:

``` python
def cross_sectional_area(target, throat_diameter='throat.diameter'):
    network = target.network
    Dt = network[throat_diameter]  # Compute area for all throats in network
    A = <input desired formula here>
    return A[network['throat.'+target.name]  # only return values for throats of target, in case of multiple domains
```

The formula you use may be something like [this](https://d77da31580fbc8944c00-52b01ccbcfe56047120eec75d9cb2cbd.ssl.cf6.rackcdn.com/0e0ad63e-afee-48de-9bd1-442142f03db0/slide47.jpg)

**Reply by Xiaozizhuo** [reply]
Thank you professor for the hint, I have tried all the functions in **op.geometry** and got good results. But I would like to inquire further about the mathematical principle of pore network generation in OpenPNM, I am trying to understand it from the point of view of mathematical formulas, but I don't seem to find detailed instructions in your team's publications, or most of them are The governing equations of the flow.

**Reply by jgostick** [reply]
Fair point.  We don't really dwell on the details of the geometrical calculations since they are considered prety standard stuff.  Can you be a bit more specific about what you're trying to accomplish?  Then maybe I can point you in the right direction.

**Reply by Xiaozizhuo** [reply]
Thank you professor for your enthusiastic reply, maybe there is something wrong with my expression. In fact, I'm currently doing a work on the fundamentals of pore network models, e.g. how to mathematically formulate the process of generating models.
